L1Trigger/L1THGCal/src/backend/HGCalHistoClusteringImpl.cc

```diff
@@ -0,0 +1,179 @@
+
+                    minDist = d;
+
+                    if ( targetSeedsEnergy.empty() ) {
+                        targetSeedsEnergy.emplace_back( iseed, seedEnergy );
+                    }
+                    else {
+                        targetSeedsEnergy.at(0).first = iseed ;
+                        targetSeedsEnergy.at(0).second = seedEnergy;
+                    }
+                }
+                
+            }
+            
+        }
+        
+        if(targetSeedsEnergy.empty()) continue;
+        //Loop over target seeds and divide up the clusters energy
+        double totalTargetSeedEnergy = 0;
+        for (auto energy: targetSeedsEnergy){
+            totalTargetSeedEnergy+=energy.second;
+        }
+    
+        for (auto energy: targetSeedsEnergy){
+
+            double seedWeight = 1;
+            if ( cluster_association_strategy_ == EnergySplit) seedWeight = energy.second/totalTargetSeedEnergy;
+            if( mapSeedMulticluster[energy.first].size()==0) {
+                mapSeedMulticluster[energy.first] = l1t::HGCalMulticluster(clu, seedWeight) ;
+            }
+            mapSeedMulticluster[energy.first].addConstituent(clu, true, seedWeight);   
+          
+        }
+        
+    }
+
+
+    for(auto mclu : mapSeedMulticluster) multiclustersTmp.emplace_back(mclu.second);
+
+    return multiclustersTmp;
+
+}
+
+
+
+
+void HGCalHistoClusteringImpl::clusterizeHisto( const std::vector<edm::Ptr<l1t::HGCalCluster>> & clustersPtrs,
+                                                     const std::vector<std::pair<GlobalPoint, double> > & seedPositionsEnergy,
+                                                     const HGCalTriggerGeometryBase & triggerGeometry,
+                                                     l1t::HGCalMulticlusterBxCollection & multiclusters)
+{
+
+
+    /* clusterize clusters around seeds */
+    std::vector<l1t::HGCalMulticluster> multiclustersTmp = clusterSeedMulticluster(clustersPtrs,seedPositionsEnergy);    
+    /* making the collection of multiclusters */
+    finalizeClusters(multiclustersTmp, multiclusters, triggerGeometry);
+
+}
+
+
+
+
+
+void
+HGCalHistoClusteringImpl::
+finalizeClusters(std::vector<l1t::HGCalMulticluster>& multiclusters_in,
+            l1t::HGCalMulticlusterBxCollection& multiclusters_out, 
+            const HGCalTriggerGeometryBase& triggerGeometry) {
+    for(auto& multicluster : multiclusters_in) {
+        // compute the eta, phi from its barycenter
+        // + pT as scalar sum of pT of constituents
+        double sumPt=multicluster.sumPt();
+
+        math::PtEtaPhiMLorentzVector multiclusterP4(  sumPt,
+                multicluster.centre().eta(),
+                multicluster.centre().phi(),
+                0. );
+        multicluster.setP4( multiclusterP4 );
+
+        if( multicluster.pt() > ptC3dThreshold_ ){
+            //compute shower shapes
+            multicluster.showerLength(shape_.showerLength(multicluster));
```

Comment:
consider combining all of this into a method of `l1t::HGCalMulticluster` that takes `shape_` as an input, in case this pattern needs to be repeated in another algo

Reply:
I would rather combine this in a method of `HGCalShowerShape` to avoid having `DataFormats/L1THGCal` depending on `L1Trigger/L1THGCal`

Reply:
That sounds acceptable.